### ISSUE TITLE
Last LSTM should have `requires_grad` as True as default

### ISSUE THREAD

**Issue by vimarshc** [fastai/fastai#2397]
Here are the observations of a small experiment I did. I wanted to verify my understanding of the inner workings of fastai. 
[The terms defined here](https://github.com/fastai/fastai/blob/master/fastai/callbacks/rnn.py) are supposed to stabilize the LSTMs in the `learn.model`. In Jeremy's lectures, he points out that when using the `fit_one_cycle` training policy we only train the last layer. 

`learn.model` is made of a `MultiBatchEncoder` with an `AWD_LSTM` inside it and a `PoolingLinearClassifier`. 

I have shown below that the `PoolingLinearClassifier` layer has `requires_grad` set to `True`. So we have a term which is supposed to be affecting the RNNs but since `requires_grad` is set to `False` that can't happen. We should be setting `freeze_to(-2)`

This unfreezes only the last `LSTM` (for which we have added the regularization term) and should give us better results. 

```
def has_params(m:nn.Module)->bool:
    "Check if `m` has at least one parameter"
    return len(list(m.parameters())) > 0


modules = [m for m in flatten_model(learn.model) if has_params(m)]

modules

[Embedding(60000, 400, padding_idx=1),
 Embedding(60000, 400, padding_idx=1),
 LSTM(400, 1152, batch_first=True),
 ParameterModule(),
 LSTM(1152, 1152, batch_first=True),
 ParameterModule(),
 LSTM(1152, 400, batch_first=True),
 ParameterModule(),
 BatchNorm1d(1200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True),
 Linear(in_features=1200, out_features=50, bias=True),
 BatchNorm1d(50, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True),
 Linear(in_features=50, out_features=2, bias=True)]
def requires_grad_bool(m:nn.Module)->Optional[bool]:
    ps = list(m.parameters())
    return ps[0].requires_grad


for it in modules:
  print(requires_grad_bool(it),it)

False Embedding(60000, 400, padding_idx=1)
False Embedding(60000, 400, padding_idx=1)
False LSTM(400, 1152, batch_first=True)
False ParameterModule()
False LSTM(1152, 1152, batch_first=True)
False ParameterModule()
False LSTM(1152, 400, batch_first=True)
False ParameterModule()
True BatchNorm1d(1200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
True Linear(in_features=1200, out_features=50, bias=True)
True BatchNorm1d(50, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
True Linear(in_features=50, out_features=2, bias=True)

learn.freeze_to(-2)

modules = [m for m in flatten_model(learn.model) if has_params(m)]

for it in modules:
  print(requires_grad_bool(it),it)

False Embedding(60000, 400, padding_idx=1)
False Embedding(60000, 400, padding_idx=1)
False LSTM(400, 1152, batch_first=True)
False ParameterModule()
False LSTM(1152, 1152, batch_first=True)
False ParameterModule()
True LSTM(1152, 400, batch_first=True)
True ParameterModule()
True BatchNorm1d(1200, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
True Linear(in_features=1200, out_features=50, bias=True)
True BatchNorm1d(50, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
True Linear(in_features=50, out_features=2, bias=True)
```

**Comment by sgugger** [comment]
I don't understand what the issue is. In the first example, everything is frozen apart from the last group of layer (which doesn't include the last LSTM, as explained in the [ULMFiT paper](https://arxiv.org/abs/1801.06146), and in the second one, everything apart from the last two groups are frozen. 
If this is just to understand the mechanism of how the library works, please use the [forum](https://forums.fast.ai/) as we keep issues for known bugs only.

**Comment by vimarshc** [comment]
Hi @sgugger , 
The documentation states that the TAR and AR regularizations have been taken from [this](https://arxiv.org/abs/1708.02182) paper. 
Now the section describing TAR and AR states: 

```
AR penalizes activations that are significantly larger than 0 as
a means of regularizing the network.
```
and 
```
TAR falls under the broad category of slowness regularizers (Hinton, 1989; Földiák, 1991; Luciw & Schmidhuber, 2012; Jonschkowski & Brock, 2015) which penalize the model from producing large changes in the hidden state.
```
which imply these terms are intended to have a regularization effect on the LSTM. 

So the mismatch as per my understanding is this: 
In its default state, we are adding a term to regularize the last LSTM in the AWD_LSTM but that can't happen because we are not letting the gradients affect the last LSTM. 
Thus, when everything is frozen apart from the last layer we should not be adding the AR and  TAR terms. The regularization term should be added when the LSTM which we are using to compute the term is not frozen.

**Comment by sgugger** [comment]
I still don't understand what the problem is. The penalty is the same kind as weight decay, which is only applied to unfrozen layers. When the network is frozen, AR/TAR is added to the loss but doesn't add anything to the gradients since the variables it depends on are not flagged with them, it only gets applied when you unfreeze those layers, exactly like weight decay isn't applied to frozen layers.

It doesn't mean we should always unfreeze all layers to have weight decay applied. The effects of gradually unfreezing have been shown to be beneficial in the ablation studies of the ULMFiT paper. AS for your experiment, yes a network frozen will have poorer results than a network with more trainable layers but that's not a fair comparison (and you will likely see the same difference with no AR/TAR). SOTA for a forward AWD-IMDB is 94.9% as shown in[this notebook](https://github.com/fastai/fastai/blob/master/examples/ULMFit.ipynb).

**Comment by vimarshc** [comment]
Hi @sgugger, 
I apologize for the lack of clarity. I am not trying to argue against the benefits of gradually unfreezing the layers. 

My only confusion is, won't adding terms to the loss which the model can't reduce have a detrimental effect on the learning process?

**Comment by sgugger** [comment]
No it won't, since the derivative of this with respect to the trainable layers will be 0. It just doesn't have any effect until we unfreeze up to the last to second layer (or more).
